src/openvic-simulation/military/Unit.cpp

```diff
@@ -235,32 +242,33 @@ bool UnitManager::add_naval_unit(std::string_view identifier, UNIT_PARAMS, NAVY_
 	return units.add_item(NavalUnit { identifier, UNIT_ARGS, NAVY_ARGS });
 }
 
-static bool shared_keys_callback(std::string_view key, ast::NodeCPtr) {
-	static const std::set<std::string, std::less<void>> reserved_keys = {
-		"icon", "type", "sprite", "active", "unit_type", "floating_flag", "priority",
-		"max_strength", "default_organisation", "maximum_speed", "weighted_value",
-		"build_time", "build_cost", "supply_consumption", "supply_cost"
-	};
-	if (reserved_keys.contains(key)) return true;
-	Logger::error("Invalid key: ", key);
-	return false;
-};
-
 bool UnitManager::load_unit_file(GoodManager const& good_manager, ast::NodeCPtr root) {
 	return expect_dictionary([this, &good_manager](std::string_view key, ast::NodeCPtr value) -> bool {
+		Unit::type_t type;
 		Unit::icon_t icon = 0;
-		std::string_view type, unit_type, sprite;
+		std::string_view unit_type, sprite, move_sound, select_sound; // TODO defaults for move_sound and select_sound
 		bool active = true, floating_flag = false;
 		uint32_t priority = 0;
 		Timespan build_time;
 		fixed_point_t maximum_speed = 0, max_strength = 0, default_organisation = 0, weighted_value = 0, supply_consumption = 0;
-		std::map<Good const*, fixed_point_t> build_cost, supply_cost;
+		Good::good_map_t build_cost, supply_cost;
 
+		using enum Unit::type_t;
+		static const string_map_t<Unit::type_t> type_map = {
+			{ "land", LAND }, { "naval", NAVAL }
+		};
+		bool ret = expect_key("type", expect_identifier(expect_mapped_string(type_map, assign_variable_callback(type))))(value);
+
+		if (!ret) {
+			Logger::error("Failed to read type for unit: ", key);
+			return false;
+		}
+
+		key_map_t key_map;
 		//shared
-		bool ret = expect_dictionary_keys_and_default(
-			key_value_success_callback,
+		ret &= add_key_map_entries(key_map,
```

Comment:
Just gonna point out you could've stored the `expect_dictionary_keys_and_default` as a callback_t and then used that stored function as the default call for the other units, would reduce the need for extra allocations with a key map

Reply:
`expect_dictionary_keys_and_default` is a `node_callback_t` while the default callbacks are `key_value_callback_t`. All the `expect_dictionary_keys` functions use `add_key_map_entries` internally, not as part of the callback but before it when the `expect_...` function is called.